### PR TITLE
feat: add all-the-icons support

### PR DESCRIPTION
**Pull request by librephoenix**
This PR adds all-the-icons support via a new variable `magit-file-icons-icon-backend`.  This defaults to `'nerd-icons`, but is automatically set to `'all-the-icons` as a fallback if nerd-icons isn't installed.  It also abstracts the icon getter functions into `magit-file-icons-icon-for-file-func` and `magit-file-icons-icon-for-dir-func` which can be set as custom function aliases if `magit-file-icons-icon-backend` is set to nil (in order to potentially support icon backends other than nerd-icons or all-the-icons).

I made just a few mistakes as I was debugging so there's some unneeded reversions in some of the commits.  If you'd like me to redo them to be more clean (or fix any other part of it), I'd be happy to.

Also, this is my first PR ever!  I hope its helpful!  Thanks for making this package!

### CODE REVIEW

**Comment by gekoke** [comment]
Hi! Thanks a lot!

> I made just a few mistakes as I was debugging so there's some unneeded reversions in some of the commits. If you'd like me to redo them to be more clean (or fix any other part of it), I'd be happy to.

No worries, I'll probably squash merge anyway.

Just a heads up, you might have to rebase onto `main` after #7 is merged, as it does modify some of the patching code that had to change due to recent updates in `magit`.

> Also, this is my first PR ever!

Congrats!

**Comment by gekoke** [comment]
On a different note, you may have noticed that I follow the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) commit message convention for this project. In the future, it's best to conform to the commit message conventions of any project you contribute to (including in the PR title, as this is used to infer commit messages in various GitHub merge operations).

**Comment by gekoke** [comment]
Without even delving into the changes themselves, we might want to consider dropping the dependency on `nerd-icons` now that we will support more than one backend, and leave it to the users to provide whichever backend they want.

File: magit-file-icons.el
Comment (gekoke) [inline magit-file-icons.el:96]: I think we can get rid of this line if we call the refresh function in the minor-mode anyway
```suggestion
```

File: magit-file-icons.el
Comment (gekoke) [inline magit-file-icons.el:70]: Why not use `:type 'function`? It's a perfectly valid `defcustom` type, and means you don't have to explicitly explain how to customize the variable in the docstring.

File: magit-file-icons.el
Comment (gekoke) [inline magit-file-icons.el:65]: I don't think this check is necessary if we have `nerd-icons` in the `Package-Requires` clause

File: magit-file-icons.el
Comment (gekoke) [inline magit-file-icons.el:89]: Don't mutate customizable variables - these are only meant to be touched by the user. Instead, define an internal variable and derive its initial state from any values for custom variables set by the user.

File: magit-file-icons.el
Comment (gekoke) [inline magit-file-icons.el:134]: I think it's better to combine these into an `or` clause, or just run the refresh unconditionally
```suggestion
    (magit-file-icons-refresh-backend)
```

File: magit-file-icons.el
Comment (gekoke) [inline magit-file-icons.el:135]: I think we can get rid of these hooks - it's probably sufficient to refresh on minor mode activation.

**Review by gekoke — CHANGES_REQUESTED** [review]

**Comment by librephoenix** [comment]
Thanks for all of the targeted feedback!  It was very helpful to learn from!

I decided to just make a new branch called `abstract-icon-getters` based on the updated main branch and I think the new implementation should be much simpler.  I'll close this PR and open a new one with the updated chages, and you'll have to let me know what you think!

Thanks!

**Pull request closed without merging**
